### PR TITLE
Support In-tree Azure Cloud Provider

### PR DESCRIPTION
**Pull request by c3y1huang**
## Why is this PR needed?

Support skuba `cloud-provider=azure`.

Ref: https://github.com/SUSE/avant-garde/issues/1582

## What does this PR do?

When provide `cpi_enable = true` in terraform.tfvars sets 
* system managed identity to all cluster virtual machines.
* assign route-table to vnet subnet

When skuba init with `cloud-provider=azure` this populates Azure cloud folder and enables cluster bootstrap with Azure in-tree cloud provider. 

## Info for QA

[New test-cases for Azure as cloud provider](https://github.com/SUSE/caasp-test-cases/pull/44)

### Related info

* [Support Azure infra deployment](https://github.com/SUSE/avant-garde/issues/1496)
* [Azure Storage Integration - for Internal Preview Only](https://confluence.suse.com/x/poInHg)

### Status **BEFORE** applying the patch

```
Flags:
      --cloud-provider string        Enable cloud provider integration with the chosen cloud. Valid values: aws, openstack, vsphere
```

### Status **AFTER** applying the patch

```
Flags:
      --cloud-provider string        Enable cloud provider integration with the chosen cloud. Valid values: aws, azure, openstack, vsphere
```

## Docs

[Add Azure cloud provider support document](https://github.com/SUSE/doc-caasp/pull/878)

# Merge restrictions

(Please do not edit this)

We are in *v4-maintenance phase*, so we will restrict what can be merged to prevent unexpected surprises:

    What can be merged (merge criteria):
        2 approvals:
            1 developer: code is fine
            1 QA: QA is fine
        there is a PR for updating documentation (or a statement that this is not needed)

Signed-off-by: Chin-Ya Huang <chin-ya.huang@suse.com>

### CODE REVIEW

File: pkg/skuba/actions/cluster/init/assets.go [hunk 1]
```diff
@@ -27,8 +27,9 @@ var (
 This directory contains all supported cloud provider integration configurations
 and instructions.
  
-- [~Openstack~](openstack/README.md)
 - [~AWS~](aws/README.md)
+- [~Azure~](aws/README.md)
```
Comment (innobead) [inline pkg/skuba/actions/cluster/init/assets.go:31]: It shoud be `azure/README.md`

**Comment by c3y1huang** [comment]
rebased.

File: ci/infra/azure/terraform.tfvars.example [hunk 2]
```diff
@@ -90,3 +90,6 @@ authorized_keys = []
 # List of Virtual Network IDs to join via virual network peering
 # EXAMPLE:`
 # peer_virutal_network_id = ["vnet-id1", "vnet-id2"]
+
+# Enable CPI integration with vSphere
```
Comment (jenting) [inline ci/infra/azure/terraform.tfvars.example:94]: vSphere?

File: ci/infra/azure/variables.tf [hunk 3]
```diff
@@ -173,3 +173,8 @@ variable "peer_virutal_network_id" {
   default     = []
   description = "IDs of a Virtual Network to connect to via a peering connection"
 }
+
+variable "cpi_enable" {
+  default     = false
+  description = "Enable CPI integration with Azure"
+}
```
Comment (jenting) [inline ci/infra/azure/variables.tf:180]: I did not saw `cpi_enable` in AWS. So I don't know what's the purpose for let users determine `cpi_enable=false` because I thought for public cloud, the CPI always on.
Could you please elaborate more.

File: pkg/skuba/actions/cluster/init/assets.go [hunk 4]
```diff
@@ -166,5 +167,52 @@ folder = "<VC_VM_FOLDER>"
...
+
+	azureReadme = strings.ReplaceAll(`# ~Azure~ integration
+
+Create a file inside this directory named ~azure.conf~, with the [supported contents](https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/docs/cloud-provider-config.md):
+
+You can find a template named azure.conf.template inside this directory.
+
+~~~
+{
+    "cloud":"AzurePublicCloud",
```
Comment (jenting) [inline pkg/skuba/actions/cluster/init/assets.go:180]: nit 
```suggestion
    "cloud": "AzurePublicCloud",
```

File: pkg/skuba/actions/cluster/init/assets.go [hunk 5]
```diff
@@ -166,5 +167,52 @@ folder = "<VC_VM_FOLDER>"
...
+
+## Important
+
+When the cloud provider integration is enabled, it's very important to bootstrap
+and join nodes with the node names same as ~Azure~ virtual machine's hostnames,
+as these names will be used by services to reconcile node metadata.
+`, "~", "`")
+
+	azureCloudConfTemplate = `{
+    "cloud":"AzurePublicCloud",
```
Comment (jenting) [inline pkg/skuba/actions/cluster/init/assets.go:205]: nit
```suggestion
    "cloud": "AzurePublicCloud",
```

File: ci/infra/azure/variables.tf (hunk 3 above)
Comment (c3y1huang) [reply ci/infra/azure/variables.tf:180]: Currently we support deployment with and without cloud provider. If all agree I can remove `cpi_enable` and consolidate the resource creation. WDYT @innobead , @cclhsu .

File: ci/infra/azure/terraform.tfvars.example (hunk 2 above)
Comment (c3y1huang) [reply ci/infra/azure/terraform.tfvars.example:94]: Ugh! thanks.

File: pkg/skuba/actions/cluster/init/assets.go (hunk 4 above)
Comment (c3y1huang) [reply pkg/skuba/actions/cluster/init/assets.go:180]: thanks

File: pkg/skuba/actions/cluster/init/assets.go (hunk 5 above)
Comment (c3y1huang) [reply pkg/skuba/actions/cluster/init/assets.go:205]: thanks.

File: ci/infra/azure/variables.tf (hunk 3 above)
Comment (innobead) [reply ci/infra/azure/variables.tf:180]: > I did not saw `cpi_enable` in AWS. So I don't know what's the purpose for let users determine `cpi_enable=false` because I thought for public cloud, the CPI always on.
> Could you please elaborate more.

A good question to rethink a bit.

There is no AWS cloud provider integration yet in AWS TFs, so this is why `cpi_enable` not there. I think we are implicitly adopting the way coming from the initial examples of `openstack` and `vsphere`, so this is why `cpi_enable` introduced here.

Let's look around the market, `rancher` provides customers to decide if setting the cloud config of a specific infra for integrating the corresponding cloud provider.  ref: https://rancher.com/docs/rke/latest/en/config-options/cloud-providers/. However, openshift 4 seems not to disclose any details about it, but 3 does the way as Rancher RKE. 

I would suggest let's keep this design to have two options for users like rancher does, and look for any feedback in the future to see if we should stick on always enabling cloud provider integration. Also, this can keep the same design as what we already provide like `openstack` and `vsphere`.

- disable cloud provider, and use on-prem like deployment on cloud
- enable cloud provider, and use cloud provider integration

Comment (jenting) [reply ci/infra/azure/variables.tf:180]: Cool, because I was thinking if _without_ cloud provider enabled in public cloud, how the user enables ingress & load balancer related function :thinking: .
But I'm lack of knowledge for this because my past experience was always enabled for the public cloud.

But as you mentioned we did not have AWS cloud provider integration yet in AWS TFs. But if I remember correctly, after AWS deployment, I could see the external IPs from `kubectl get nodes -o wide`. So I'm not sure if AWS TFs by default is _cpi_enable=true_ or _cpi_enable=false_.

I agree to keep current design but just curious what's the default value `cpi_enable` should be.

Comment (innobead) [reply ci/infra/azure/variables.tf:180]: Yes, you are right. native ingress integration is provided by Azure cloud provider integration, ref: https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/docs/services/README.md.

After discussing with @jenting again, some conclusions to follow up later on.
- deployment: skuba provides optionally enabling cloud provider (as what we have no by providing `--cloud-provider`)
- infra: should we have `cpi_enable` in TFs, or we should have agnostic terraforms scripts regardless if having cloud provider integration or not. To have better user experience. 

To sum up the current cloud providers we have right now. 
- `openstack`: `cpi_enable` is used to automatically generate and copy openstack.conf to each node.
- `vsphere`: cpi_enable` is used to manage `vsphere_folder` resource
- `azure`: @c3y1huang is still working on that. 

@c3y1huang let's see if it's possible to remove `cpi_enable`, and it can be working regardless `--cloud-provider` enabled or not. If yes, this always achieves what we discussed.

And, later we definitely need to revisits `vsphere` and `openstack` to make them consistent behavior.

**Comment by c3y1huang** [comment]
Verified managed identity:
* file storage, disk storage; 
* in-line volume, persistent volume, storage class.

**Comment by c3y1huang** [comment]
Verified sp secret:

* single node of each type, multi nodes
* file storage, disk storage;
* in-line volume, persistent volume, storage class.

File: ci/infra/azure/README.md [hunk 6]
```diff
@@ -9,11 +9,25 @@ to deploy CaaS Platform on top of it.
...
+
+You can also use Azure CLI to automate the task. For installation instructions, refer to link:https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest[Install the Azure CLI].
+```bash
+az ad sp create-for-rbac --role="Contributor" --scopes="/subscriptions/<SUBSCRIPTION_ID>"
+```
+
+Update and source `container-openrc.sh` before deploying terraform script.
+```bash
+source container-openrc.sh
+```
```
Comment (innobead) [inline ci/infra/azure/README.md:28]: Is this for Azure? it seems like for `openstack`?

File: ci/infra/azure/README.md [hunk 7]
```diff
@@ -40,11 +54,7 @@ the [Azure Linux Agent](https://docs.microsoft.com/en-us/azure/virtual-machines/
 
 ### Using spot instances
 
-It's possible to create a [spot VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/spot-vms)
-both for the master and worker nodes.
-
-This can be done by setting these variables to `true` (they are set to
-`false` by default):
+User can enable nodes to use [spot VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/spot-vms) by setting below to `true` (`false` by default).
```
Comment (innobead) [inline ci/infra/azure/README.md:57]:
```suggestion
User can enable nodes to use [spot VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/spot-vms) by setting below variables to `true` (`false` by default).
```

File: ci/infra/azure/README.md [hunk 8]
```diff
@@ -157,3 +167,17 @@ It is possible to join existing network to the cluster.  It can be setup by addi
 It is possible to enable multiple zone.  It can be set `enable_zone` to `true` and master/worker node will distribute sequentially based on zones defined in `azure_availability_zones`.
 
   * As of June 2020, the [Azure Availability Zones](https://docs.microsoft.com/en-us/azure/availability-zones/az-region) is not available in all Azure regions.
+
+#enable-cloud-provider-interface
+## Enable Cloud Provider Interface
+
+Configure `cpi_enable` to `true` enables terraform to provision the cluster with setups for cloud provider interface.
```
Comment (innobead) [inline ci/infra/azure/README.md:174]:
```suggestion
Configuring `cpi_enable` to `true` enables terraform to provision the cluster with setups for the cloud provider interface.
```

File: ci/infra/azure/network.tf [hunk 9]
```diff
@@ -66,3 +66,22 @@ resource "azurerm_private_dns_zone_virtual_network_link" "internal_dns" {
   virtual_network_id    = azurerm_virtual_network.virtual_network.id
   registration_enabled  = true
 }
+
+# Route Table
+resource "azurerm_route_table" "nodes" {
```
Comment (innobead) [inline ci/infra/azure/network.tf:71]: Are route resources here for managing routing rules by azure cloud provider?

File: ci/infra/azure/README.md (hunk 6 above)
Comment (c3y1huang) [reply ci/infra/azure/README.md:28]: Yes this is for Azure. I believe the idea did came from openstack, right @cclhsu ?
ref: https://github.com/SUSE/skuba/blob/master/ci/infra/azure/container-openrc.sh

File: ci/infra/azure/network.tf (hunk 9 above)
Comment (c3y1huang) [reply ci/infra/azure/network.tf:71]: Yes need to exist for the CCM.

File: ci/infra/azure/README.md (hunk 6 above)
Comment (innobead) [reply ci/infra/azure/README.md:28]: Do we really need this? @cclhsu I think we already use the service principal to do the authentication instead of environment variables.

Comment (c3y1huang) [reply ci/infra/azure/README.md:28]: There are two possible ways for terraform takes credential configs:
* environment variable (recommended) <- where we are
* in-line
ref: https://www.terraform.io/docs/providers/azurerm/guides/service_principal_client_secret.html

Comment (innobead) [reply ci/infra/azure/README.md:28]: Yes, I misread here :(, the thing is correct as what we decided to do. Thanks for clarifying. All good now thanks @c3y1huang and @cclhsu .

Comment (cclhsu) [reply ci/infra/azure/README.md:28]: Yes.  Same method is used for openstack and vsphere, and the question was also raised during demo (I think for vshpere) if this way is possible.

Comment (innobead) [reply ci/infra/azure/README.md:28]: Thanks for clarifying @cclhsu :+1:

**Comment by c3y1huang** [comment]
rebased.

File: pkg/skuba/actions/cluster/init/assets.go [hunk 10]
```diff
@@ -27,8 +27,9 @@ var (
 This directory contains all supported cloud provider integration configurations
 and instructions.
  
-- [~Openstack~](openstack/README.md)
 - [~AWS~](aws/README.md)
+- [~Azure~](azure/README.md)
+- [~Openstack~](openstack/README.md)
 - [~vSphere~](vsphere/README.md)
```
Comment (jenting) [inline pkg/skuba/actions/cluster/init/assets.go:34]: Thanks for the sorting :heart:

File: ci/infra/azure/master-instance.tf [hunk 11]
```diff
@@ -70,3 +77,16 @@ resource "azurerm_virtual_machine_extension" "master" {
     }
 SETTINGS
 }
+
+locals {
+  master_principal_ids = var.cpi_enable ? azurerm_linux_virtual_machine.master.*.identity.0.principal_id : []
+}
+
+resource "azurerm_role_assignment" "master" {
```
Comment (pgonin) [inline ci/infra/azure/master-instance.tf:85]: What role are you assigning to the masters here ?

It should be an option to manage this with terraform, as it requires the Operator to have full control over the subscription, which is not an ideal security model (at least we wanted not QA engineers to have 'Owner' role over SUSE Engineering subscription as a concrete example)
Instead documentation should explain how the owner could provide the required configuration.

**Comment by maximenoel8** [comment]
### QA information

#### Configuration
###### skuba
 - build from skuba / devel

###### Deployment 
- deploy from airgapped server with devel container images and packages
- zone is disable during deployment ( `"enable_zone": false,`) 
- lb / 2 master / 3 workers

###### Storage class 

```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: sample-disk-sc
  annotations:
    storageclass.kubernetes.io/is-default-class: "true"
provisioner: kubernetes.io/azure-disk
parameters:
  skuName: StandardSSD_LRS
  location: "West Europe"
```

#### Results

I was able to create pvc and deploy services using pvc.

**Comment by maximenoel8** [comment]
### QA information

#### Configuration
###### skuba
 - build from skuba / devel

###### Deployment 
- deploy from airgapped server with devel container images and packages
- zone is enable during deployment ( `"enable_zone": true,` and `"azure_availability_zones": ["1","2","3"],`) 
- lb / 2 master / 3 workers

###### Storage class 

```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: sample-disk-sc
  annotations:
    storageclass.kubernetes.io/is-default-class: "true"
provisioner: kubernetes.io/azure-disk
parameters:
  skuName: StandardSSD_LRS
  location: "West Europe"
```

#### Results

Using the storage class is failing.


#### Error 
```
RetryAfter: 0s, HTTPStatusCode: 400, RawError: {
  "error": {
    "code": "BadRequest",
    "message": "Disk /subscriptions/c011786b-59d7-4817-880c-7cd8a6ca4b19/resourceGroups/mnoel-cluster-x8rc-1-resource-group/providers/Microsoft.Compute/disks/cluster-dynamic-pvc-a63a9428-26fb-47a7-b96c-c7409b86b938 cannot be attached to the VM because it is not in zone '2'."
  }
}
```

**Comment by innobead** [comment]
> ### QA information
> #### Configuration
> ###### skuba
> * build from skuba / devel
> 
> ###### Deployment
> * deploy from airgapped server with devel container images and packages
> * zone is enable during deployment ( `"enable_zone": true,` and `"azure_availability_zones": ["1","2","3"],`)
> * lb / 2 master / 3 workers
> 
> ###### Storage class
> ```
> kind: StorageClass
> apiVersion: storage.k8s.io/v1
> metadata:
>   name: sample-disk-sc
>   annotations:
>     storageclass.kubernetes.io/is-default-class: "true"
> provisioner: kubernetes.io/azure-disk
> parameters:
>   skuName: StandardSSD_LRS
>   location: "West Europe"
> ```
> 
> #### Results
> Using the storage class is failing.
> 
> #### Error
> ```
> RetryAfter: 0s, HTTPStatusCode: 400, RawError: {
>   "error": {
>     "code": "BadRequest",
>     "message": "Disk /subscriptions/c011786b-59d7-4817-880c-7cd8a6ca4b19/resourceGroups/mnoel-cluster-x8rc-1-resource-group/providers/Microsoft.Compute/disks/cluster-dynamic-pvc-a63a9428-26fb-47a7-b96c-c7409b86b938 cannot be attached to the VM because it is not in zone '2'."
>   }
> }
> ```

@maximenoel8 please help create an issue for this failed case. Thanks for the regression :+1: 

cc @c3y1huang @cclhsu @jenting
